apps/dashr-ocean-optics/app.R

```diff
@@ -0,0 +1,580 @@
+      list('l2')
+    }
+  }
+)
+
+
+# update the plot
+app$callback(
+  output = list(id = "spec-readings", property = "figure"),
+  params = list(
+    input(id = "spec-reading-interval", property = "n_intervals"),
+    state(id = "power-button", property = "on"),
+    state(id = "autoscale-switch", property = "on"),
+    state(id = "light-intensity-knob", property = "value"),
+    state(id = "hidden-div-int-time", property = "children"),
+    state(id = "hidden-div-light-source", property = "children")
+  ),
+
+  function(n, pwr, auto_range, knob_intensity, scale, ls) {
+
+    ls <- unlist(ls)
+    knobIntensity <- ifelse(ls == "l2", as.numeric(unlist(knob_intensity)), 0)
+
+    scale <- as.numeric(unlist(scale))
+
+    # intensity computation lags behind when `length.out` is more than 1000!
+    wavelengths <- seq(400, 900, length.out = 5000L)
+
+    if (pwr) {
+      intensities <- SampleSpectrum(scale = scale, knobIntensity = knobIntensity, wavelengths)
+    } else {
+      intensities <- rep(0, length(wavelengths))
+    }
+
+    # start creating layout variables
+    xAxis <- list(
+      'title' = 'Wavelength (nm)',
+      'titlefont' = list(
+        'family' = 'Helvetica, sans-serif',
+        'color' = colors[['secondary']]
+      ),
+      'tickfont' = list(
+        'color' = colors[['tertiary']]
+      ),
+      'dtick' = 100,
+      'color' = colors[['secondary']],
+      'gridcolor' = colors[['grid-colour']]
+    )
+
+    yAxis <- list(
+      'title' = 'Intensity (AU)',
+      'titlefont' = list(
+        'family' = 'Helvetica, sans-serif',
+        'color' = colors[['secondary']]
+      ),
+      'tickfont' = list(
+        'color' = colors[['tertiary']]
+      ),
+      'color' = colors[['secondary']],
+      'gridcolor' = colors[['grid-colour']]
+    )
+
+    # auto-scale feature
+    if (pwr && auto_range) {
+      xAxis[['range']] <- list(min(wavelengths),
+                               max(wavelengths))
+      yAxis[['range']] <- list(min(intensities),
+                               max(intensities))
+
+    }
+
+    layout <- list(
+      height = 600,
+      font = list('family' = 'Helvetica Neue, sans-serif',
+                  'size' = 12),
+      margin = list('t' = 20),
+      titlefont = list(
+        'family' = 'Helvetica, sans-serif',
+        'color' = colors[['primary']],
+        'size' = 26),
+      xaxis = xAxis,
+      yaxis = yAxis,
+      paper_bgcolor = colors[['background']],
+      plot_bgcolor = colors[['background']]
+      )
+
+    return(
+      figure <- list(data = list(
+        list(
+          x = as.list(wavelengths),
+          y = as.list(intensities),
+          name = 'Spectrometer readings',
+          mode = 'lines',
+          line = list('width' = 1, 'color' = colors[['accent']])
+        )
+      ),
+      layout = layout)
+    )
+  }
+)
+
+if (!appName == ""){
+  app$run_server(host = "0.0.0.0", port = Sys.getenv('PORT', 8050))
+} else {
+  app$run_server(debug = TRUE)
+}
```

Comment:
If you're using RStudio, would you mind enabling newlines at EOF in the preferences? It will make the visual warning in GitHub that currently appears go away 🙂  I'd tick the second box as well, since it will save you from linting errors if you edit Python in RStudio too.

![Screenshot 2019-07-22 19 10 29](https://user-images.githubusercontent.com/9809798/61671242-b60e6c00-acb4-11e9-9a6f-8f6e346a2d4f.png)

If you're using another editor, there's likely to be a similar setting there too.